File: ru-olymp-train-winter-2007/Problems/20070119/problems/binsearch/unit-tour/pas/binsearch.pas
unit binsearch;

interface

function getN: longint;

function query(i: longint): boolean;

procedure answer(i: longint);

implementation

var
	n: longint;

function getN: longint;
begin
	repeat
    	write('enter n: ');
    	read(n);
    	if (n < 1) or (n > 60) then begin
    		writeln('invalid n, n must be between 1 and 60');
    	end;
    until (1 <= n) and (n <= 60);
    getN := n;
end;

function query(i: longint): boolean;
var
	t: longint;
begin
	if (i < 1) or (i > n) then begin
		writeln('invalid query: ', i);
		halt(1);
	end;

	repeat
		writeln('is y <= x[', i, '] (1 - yes, 0 - no)?');
		read(t);
		if (t <> 0) and (t <> 1) then
			writeln('enter 1 or 0');
	until (t = 0) or (t = 1);
	query := t = 1;
end;

procedure answer(i: longint);
begin
	writeln('program answers ', i);
end;

end.

File: ru-olymp-train-winter-2007/Problems/20070119/problems/binsearch/unit-tour/c/binsearch.c
#include <stdio.h>
#include <stdlib.h>

static int n;


int getN() {
	do {
		printf("enter n: ");
		scanf("%d", &n);
		if (n < 1 || n > 60) {
			printf("invalid n, n must be between 1 and 60\n");
		}
	} while (n < 1 || n > 60);
	return n;
}

int query(int i) {
	int t;

	if (i < 1 || i > n) {
		printf("invalid query: %d\n", i);
		exit(1);
	}

	do {
		printf("is y <= x[%d] (1 - yes, 0 - no)?\n", i);
		scanf("%d", &t);
		if (t != 0 && t != 1) {
			printf("enter 1 or 0\n");
		}
	} while (t != 0 && t != 1);
	return t;
}

void answer(int i) {
	printf("program answers %d\n", i);
}
